kpops/utils/gen_schema.py

```diff
@@ -0,0 +1,137 @@
+import logging
+from enum import Enum
+from typing import Annotated, Any, Literal, Sequence, Union, get_args, get_origin
+
+from pydantic import Field, schema, schema_json_of
+from pydantic.fields import ModelField
+from pydantic.schema import SkipField
+from rich.console import Console
+
+from kpops.cli.pipeline_config import PipelineConfig
+from kpops.cli.registry import _find_classes
+from kpops.components.base_components.pipeline_component import PipelineComponent
+
+
+class SchemaScope(str, Enum):
+    pipeline = "pipeline"
+    config = "config"
+
+
+original_field_schema = schema.field_schema
+
+
+# adapted from https://github.com/tiangolo/fastapi/issues/1378#issuecomment-764966955
+def field_schema(field: ModelField, **kwargs: Any) -> Any:
+    if field.field_info.extra.get("hidden_from_schema"):
+        raise SkipField(f"{field.name} field is being hidden")
+    else:
+        return original_field_schema(field, **kwargs)
+
+
+schema.field_schema = field_schema
+
+log = logging.getLogger("")
+
+ModelField
+
+
+def is_valid_component(
+    used_fields: list[str], component: type[PipelineComponent]
+) -> bool:
+    """
+    Check whether a PipelineComponent subclass has a valid definition for the schema generation.
+
+    :param used_fields: types defined so far
+    :type used_fields: list[str]
+    :param component: component type to be validated
+    :type component: type[PipelineComponent]
+    :return: Whether component is valid for schema generation
+    :rtype: bool
+    """
+    component_name: str = component.__name__
+    # PipelineComponent always has a field 'type'
+    type: str = component.__fields__.get("type").field_info.default  # type: ignore[union-attr]
+    if not component.__fields__.get("schema_type"):
+        log.warning(f"SKIPPED {component_name}, schema_type is not defined")
+        return False
+    schema_type: ModelField = component.__fields__.get("schema_type")  # type: ignore[assignment]
```

Comment:
you're already checking if the `schema_type` attribute exists in the condition above. here you can directly access it then: `component.__fields__["schema_type"]`

Reply:
even better would be to rely on type narrowing

example
```python
schema_type = component.__fields__.get("schema_type") # ModelField | None
if not schema_type:
    return
schema_type # ModelField
```

Reply:
done